lib/ukschedcoop/schedcoop.c

```diff
@@ -168,8 +171,11 @@ static void schedcoop_thread_woken(struct uk_sched *s, struct uk_thread *t)
 
 	if (t->wakeup_time > 0)
 		UK_TAILQ_REMOVE(&c->sleep_queue, t, queue);
-	if (t != uk_thread_current() && uk_thread_is_runnable(t))
```

Comment:
Hey @mariasfiraiala and @eduardvintila, Why don't you remove the check `t != uk_thread_current()` since if the thread is not runnable, that means wait has already called uk_thread_set_block() and  uk_sched_thread_block() . This means we can safely remove current thread from sleep queue and add it to run queue. 

So Even if we receive the interrupt `ukplat_lcpu_restore_irqf()` and `uk_sched_yield()`, we are sure that the current thread is on sleep queue and we can remove it. 

ideally if we do this, we do not need queuable flag.

Reply:
Hello, @Sairajkodilkar. By eliminating the `t != uk_thread_current()` condition, we risk placing the current thread on the `run_queue`  before the execution of `yield()`. In our case, consider the `__wq_wait_event_deadline` macro from [lib/uksched/include/uk/wait.h,](https://github.com/unikraft/unikraft/blob/staging/lib/uksched/include/uk/wait.h#L119) which is used for blocking the main thread, for example.

If an interrupt fires at this point, right before the `yield()`:
```C
            ukplat_lcpu_restore_irqf(flags); \
            if (lock) \
                unlock_fn(lock); \
            uk_sched_yield(); \
```
... and wakes up the thread by adding it to the `run_queue`, then during the yield, this assertion will fail: 
```C
    next = UK_TAILQ_FIRST(&c->run_queue);
    if (next) {
        UK_ASSERT(next != prev);
 ```
which can be seen during booting:
```console
[    0.156197] CRIT: [libukschedcoop] <schedcoop.c @   93> Assertion failure: next != prev
[    0.156494] Info: [libkvmplat] <shutdown.c @   35> Unikraft halted
```

However, thank you nevertheless for your suggestion! Eventually I have replaced the check you have mentioned with `uk_thread_is_queueable`, which is more general in our case.